### PR TITLE
Framework: Run i18n-mixin-to-localize codemod

### PR DESCRIPTION
**Pull request by ockham**
TODO:
* [x] Look for `prototype.translate` in tests, remove
* [x] Fix mixins ("headless" components still using `this.translate`)
  * [x] `sites-list` notices: see #18625
  * [x] `form-base`: Turn into component, have consumers inherit?
    * Looks okay, all consumers are `localize()`d now, so `form-base` should be able to just use `this.props.translate()`.
* [x] Check `ref`s. Do we need `localize` to hoist them?
  * Looks okay, see https://github.com/Automattic/wp-calypso/pull/18590#issuecomment-335391380 and convo below
* [x] Grep for files containing `this.props.translate` but not `i18n-calypso`. Any takers?

### CODE REVIEW

**Comment by psealock** [comment]
>`Revert PostEditor changes`

@ockham Do you know why these tests were failing by chance? It seemed a reference to the Component from `ReactDOM.render()` was working sometimes, but not others.

**Comment by ockham** [comment]
> @ockham Do you know why these tests were failing by chance? It seemed a reference to the Component from `ReactDOM.render()` was working sometimes, but not others. 

I think there are two errors in play here -- 6dc6a62 was just to fix tests; turns out the `default` export from that file was already being `localize`; the codemod wrapped the unlocalized component in a `localize` call, which broken tests

**Comment by ockham** [comment]
The other error I'm currently seeing with PostEditor is this:

image

The immediate reason for this is that `sections.js` (as generated by `server/bundler/loader`) tries to access the `isRetry` static method from `LoadingError` (in `client/layout/error`), but that component is now wrapped in a `localize`. To mitigate that, one option would be to have `localize` hoist statics, much like `react-redux`' `connect()` [does](https://github.com/reactjs/react-redux/blob/master/docs/api.md#static-methods). Or maybe we can rewrite `server/bundler/loader` and/or `client/layout/error` to decouple the `isRetry` logic from the component.

This still doesn't fix the underlying error with the editor, but maybe we'll get more meaningful error messages to track it down once we fix the `LoadingError` issue.

**Comment by ockham** [comment]
Fix in c17bc44. I think I'll spin up a separete PR for that.

**Comment by ockham** [comment]
So the actual error in PostEditor is 

image

and I think the reason are once again static methods -- https://github.com/Automattic/wp-calypso/blob/6e86844bec197936dced31392d9c9a27dd2ce393/client/components/tinymce/plugins/wpcom-view/views/contact-form/index.jsx#L20 in this case. After the codemod is run, that component is wrapped in `localize`. So we might have to modify `i18n-calypso` to hoist static methods after all.

**Comment by ockham** [comment]
Oh, we might work around that using a similar pattern as https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/wpcom-view/views/simple-payments/index.jsx

**Comment by psealock** [comment]
>Oh, we might work around that using a similar pattern as https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/wpcom-view/views/simple-payments/index.jsx

I like this pattern better because it signals intention better than independently exported functions. However, fixing the root cause of the issue by addressing the `localize` function itself is the best solution.

**Comment by sirreal** [comment]
> Grep for files containing this.props.translate but not i18n-calypso. Any takers?

```
$ ack -L 'i18n-calypso' $(ack -l 'this\.props\.(translate|moment|numberFormat)' client)

client/blocks/global-notice/README.md
client/my-sites/checkout/cart/cart-messages-mixin.jsx
client/my-sites/jetpack-manage-error-page/README.md
client/my-sites/plugins/plugin-action/README.md
client/my-sites/plugins/plugins-list/README.md
client/my-sites/sharing/connections/services/eventbrite.js
client/my-sites/sharing/connections/services/google-photos.js
client/my-sites/sharing/connections/services/instagram.js
```

I'll have a look at these… _cracks knuckles_

**Comment by sirreal** [comment]
After some grepping, all the usage seems fine. The `cart-messages-mixin` could have some extra protection, but that's probably not necessary. See #18651 for more details.

**Comment by sirreal** [comment]
> Check `ref`s. Do we need `localize` to hoist them?

What's the concern here? Here are some findings that may be relevant:

`ref`s in components that need codemodding:

```
$ ack -l '\Wref=' $(ack --js -l 'this\.(translate|moment|numberFormat)' client )

client/blocks/author-selector/index.jsx
client/components/plans/premium-popover/index.jsx
client/components/tinymce/plugins/contact-form/dialog/field-remove-button.jsx
client/components/tinymce/plugins/wplink/dialog.jsx
client/me/profile-links/add-buttons.jsx
client/me/security-2fa-backup-codes-list/index.jsx
client/me/security-2fa-sms-settings/index.jsx
client/me/security-account-recovery/edit-email.jsx
client/my-sites/domains/components/form/country-select.jsx
client/my-sites/exporter/export-card/post-type-options.jsx
client/my-sites/importer/uploading-pane.jsx
client/my-sites/media-library/header.jsx
client/my-sites/people/team-list/index.jsx
client/my-sites/people/viewers-list/index.jsx
client/my-sites/plugins/plugin-sections/index.jsx
client/my-sites/plugins/plugin-site-update-indicator/index.jsx
client/my-sites/post-selector/selector.jsx
client/my-sites/sharing/buttons/preview-buttons.jsx
client/my-sites/stats/post-trends/day.jsx
client/my-sites/stats/post-trends/index.jsx
client/post-editor/editor-status-label/index.jsx
client/post-editor/editor-sticky/index.jsx
client/post-editor/media-modal/gallery-help.jsx
client/post-editor/post-editor.jsx
```

There appear to be 84 uses of `ref` in components with `localize`. Here's an example:

https://github.com/Automattic/wp-calypso/blob/ce892aade62ea56b5942574c34cc21bafdc505ae/client/blocks/app-banner/index.jsx#L160-L163

Is verifying that that example works fine enough?

**Comment by ockham** [comment]
> What's the concern here?

String `ref`s that are referred to from outside the component, the `localize()`-wrapped component won't have the `ref` available.

> Is verifying that that example works fine enough?

Yeah, but we'd have to do that on a case-by-case basis.

The (quicker, more reliable) alternative would be to add  `ref`-hoisting to `localize()` (as `react-redux`'s `connect()` does AFAIK). I'm a bit reluctant (read: lazy) to do the whole version bumping/releasing dance around `i18n-calypso`, but it's probably the safest way.

It sucks all the more since string `ref`s are deprecated (in favor of callback `ref`s) anyway (tho they aren't going away with React 16 just yet); plus some of those string `ref`s were just used by tests AFAIK (which we could resolve by using different/enzyme-provided selectors).

Going thru `ref`s manually is obviously a more manual process, so we need to decide if it'd be blocking us for too long, and if we should just go with adding `ref` hoisting instead. (If we'd like to pursue the "manual" strategy, e2e tests might be helpful in quickly tracking down breakage; but we might want to get rid of the remaining `this.translate` calls, i.e. tackle the `form-base` stuff. first.)

**Comment by sirreal** [comment]
Good thinking!

How about this, files containing `.refs.` but _not_ `this.refs.`:

```
$ ack -L 'this\.refs\.' $(ack -l '\.refs\.' client)

client/components/drop-zone/test/index.jsx
client/components/root-child/test/index.jsx
client/components/segmented-control/index.jsx
client/devdocs/test/doc.jsx
client/my-sites/domains/domain-management/list/test/index.js
client/my-sites/themes/themes-magic-search-card/index.jsx
```

Looks like a manageable list, I'll dig further…

**Comment by sirreal** [comment]
After looking a bit, cleaning up all those `ref`s looks like it might get a bit hairy. Note the especially fun ones like `this.refs[ 'item-' + newIndex ].refs.itemLink`.

@ockham Let's hear what you think, but it may be best to add the `ref` hoisting to `localize`. Sounds like it should really be there anyways (at least until string `ref`s [move from "leagacy" to "deprecated"](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs)).

Here's the full output of non-`this` refs:

```
$ ack --js '(?<!this)\.refs' client

client/components/drop-zone/test/index.jsx
60:             expect( tree.refs.zone.parentNode.id ).to.equal( 'container' );
72:             expect( tree.refs.zone.parentNode.id ).to.not.equal( 'container' );
73:             expect( tree.refs.zone.parentNode.parentNode ).to.eql( document.body );

client/components/root-child/test/index.jsx
55:                     expect( tree.refs.parentChild.parentNode.className ).to.equal( 'parent' );
57:                     expect( tree.refs.rootChild.parentNode.parentNode ).to.eql( document.body );
68:                     expect( tree.refs.rootChild.parentNode.className ).to.equal( 'wrapper' );
70:                     expect( tree.refs.rootChild.parentNode.parentNode.parentNode ).to.eql( document.body );

client/components/segmented-control/index.jsx
221:            ReactDom.findDOMNode( this.refs[ 'item-' + newIndex ].refs.itemLink ).focus();

client/components/select-dropdown/index.jsx
171:                                                    self.refs.dropdownContainer.focus();
401:            ReactDom.findDOMNode( this.refs[ 'item-' + newIndex ].refs.itemLink ).focus();

client/devdocs/test/doc.jsx
36:                             const html = ReactDom.findDOMNode( renderedSingleDoc.refs.body ).innerHTML;

client/extensions/wp-super-cache/components/advanced/directly-cached-files.jsx
27:             const newDirectPage = this.refs.newDirectPage.refs.textField;

client/my-sites/domains/domain-management/list/test/index.js
122:                            assert( ReactDom.findDOMNode( component.refs.cancelChangePrimaryButton ) );
126:                            const button = ReactDom.findDOMNode( component.refs.cancelChangePrimaryButton );

client/my-sites/themes/themes-magic-search-card/index.jsx
102:            const { value, selectionStart } = this.refs[ 'url-search' ].refs.searchInput;
142:                            cursorPosition: val.slice( 0, this.refs[ 'url-search' ].refs.searchInput.selectionStart )
```

**Comment by ockham** [comment]
Oh wow, thanks a lot for the grepping, really helps to see the bigger picture here!

I took a random sample from that list: `client/components/root-child/test/index.jsx` really only uses the `ref` for testing, and the tested component was already `localize()`d before (might be the case for a few other ones, too). I might take some more samples, but I'm kinda adamant to cross fingers and discover the problematic ones in testing. I'm a bit hesitant to add `ref` hoisting to `i18n-calypso` when string `ref`s are on their way out...

**Comment by ockham** [comment]
Other instances are textbook cases of [`ref` function attributes](https://reactjs.org/docs/refs-and-the-dom.html#adding-a-ref-to-a-dom-element), e.g. `client/components/segmented-control/index.jsx`...

**Comment by ockham** [comment]
> Here's the full output of non-`this` refs: [...]

I had a look at the rest of the list. Tests should be fine by definition (they wouldn't pass otherwise), and the other ones seem to either use `localize()` already, or aren't using `translate()` at all.

**Comment by ockham** [comment]
I just vetted `form-base`, see note in PR desc. Seems good. Thus: 368d5c0

**Comment by ockham** [comment]
Is it me or does this thing feel way faster now that the auto-injected mixin is gone? 😮

**Comment by rachelmcr** [comment]
I ran through our full e2e tests on this branch and everything looks good. 👍

**Comment by samouri** [comment]
Nice work @ockham!!! This was a tricky one

File: client/blocks/author-selector/index.jsx [hunk 1]
```diff
@@ -241,9 +242,9 @@ const SwitcherShell = React.createClass( {
 	_onSearch: function( searchTerm ) {
 		this.props.updateSearch( searchTerm );
 	},
-} );
+} ));
 
-module.exports = React.createClass( {
+module.exports = localize(React.createClass( {
```
Comment (spen) [inline client/blocks/author-selector/index.jsx]: Related: #18710 will soon be ready for review (once a desc is written).
It aims to fix up instances similar to this ^

It currently only looks at `export default`s but could be tweaked to consider `module.exports =` too :)

**Pull request merged**